### ISSUE TITLE
Issues with 2.5.2 (dumping results, committing transactions, and timeout behaviors)

### ISSUE THREAD

**Issue by noureldin-khaled** [neo4j-php/neo4j-php-client#124]
Hello :wave: 

So I have upgraded to version 2.5.2 to be able to use the timeout fix reported here https://github.com/neo4j-php/neo4j-php-client/issues/102#issuecomment-1055904982
But I noticed some other issues happening and wanted to list them down here

All issues create the client as follows: (with varying timeout value)
```php
$url = 'bolt://test:test@neo4j:7687?database=test'
$this->client = ClientBuilder::create()
                             ->withDriver('bolt', $url, null)
                             ->withDefaultTransactionConfiguration(TransactionConfiguration::default()->withTimeout(150.0))
                             ->withDefaultDriver('bolt')
                             ->build();
```

**Dumping a query result**
I run a query and then want to dump the result to see it, but an error occurs
Steps:
1. Create client
2. Run a query then dump the results
```php
$result = $client->run("MATCH (n:Node) SET n.testing = 'hello' RETURN n");
dump($result);
```
This error is produced
```
In VarCloner.php line 108:
                                                                                                                                                                                                 
  Type error: Cannot assign Symfony\Component\VarDumper\Cloner\Stub to reference held by property Laudis\Neo4j\Databags\SummarizedResult::$summary of type ?Laudis\Neo4j\Databags\ResultSummary
```
This happens for both transactions and direct runs.
If I dump `dump($result->getSummary());` it works and prints the summary, is this an intended behavior? I know this is not a critical issue, but I thought I would mention it :sweat_smile: 

Note: something weird happens with the run, if a result is expected to be return (there is a return in the query), just the run itself doesn't run the query, but `getSummary()` or `toArray()` do, so this
```php
$result = $client->run("MATCH (n:Node) WITH * CALL apoc.util.sleep(2000000) RETURN n");
```
finishes instantly, indicating that the query is not being ran or something.
But this
```php
$result = $client->run("MATCH (n:Node) WITH * CALL apoc.util.sleep(2000000) RETURN n");
$summary = $result->getSummary();
```
or this
```php
$result = $client->run("MATCH (n:Address) WITH * CALL apoc.util.sleep(2000000) RETURN n");
$array = $result->toArray();
```
Actually run the query and perform the sleep properly, is this behavior intended?

**Committing a transaction causes an error**
This issue is similar to https://github.com/neo4j-php/neo4j-php-client/issues/117
Run a query through a transaction, then commit, an error occurs indicating a failure on the DISCARD message
Steps:
1. Create client
2. Run a query through a transaction then commit
```php
$tsx = $client->beginTransaction();
$result = $tsx->run("MATCH (n:Node) SET n.testing = 'world' RETURN n");
$tsx->commit();
```
This error is produced
```
In V4.php line 94:
                                                                                                                                             
  Message 'DISCARD Map{qid -> Long(0), n -> Long(-1)}' cannot be handled by a session in the READY state. (Neo.ClientError.Request.Invalid) 
```
So I check the debug logs in neo4j and found this error
```
2022-03-16 17:34:35.029+0000 DEBUG [o.n.b.v.m.BoltResponseMessageWriterV3] Encoding a fatal failure message to send. Message: FAILURE Invalid Message 'DISCARD Map{qid -> Long(0), n -> Long(-1)}' cannot be handled by a session in the READY state.
2022-03-16 17:34:35.029+0000 ERROR [o.n.b.r.DefaultBoltConnection] Protocol breach detected in bolt session 'bolt-20'.
org.neo4j.bolt.runtime.BoltProtocolBreachFatality: Message 'DISCARD Map{qid -> Long(0), n -> Long(-1)}' cannot be handled by a session in the READY state.
	at org.neo4j.bolt.runtime.statemachine.impl.AbstractBoltStateMachine.nextState(AbstractBoltStateMachine.java:159) ~[neo4j-bolt-4.4.3.jar:4.4.3]
	at org.neo4j.bolt.runtime.statemachine.impl.AbstractBoltStateMachine.process(AbstractBoltStateMachine.java:102) ~[neo4j-bolt-4.4.3.jar:4.4.3]
	at org.neo4j.bolt.messaging.BoltRequestMessageReader.lambda$doRead$1(BoltRequestMessageReader.java:93) ~[neo4j-bolt-4.4.3.jar:4.4.3]
	at org.neo4j.bolt.runtime.DefaultBoltConnection.lambda$enqueue$0(DefaultBoltConnection.java:156) ~[neo4j-bolt-4.4.3.jar:4.4.3]
	at org.neo4j.bolt.runtime.DefaultBoltConnection.processNextBatchInternal(DefaultBoltConnection.java:252) ~[neo4j-bolt-4.4.3.jar:4.4.3]
	at org.neo4j.bolt.runtime.DefaultBoltConnection.processNextBatch(DefaultBoltConnection.java:187) ~[neo4j-bolt-4.4.3.jar:4.4.3]
	at org.neo4j.bolt.runtime.DefaultBoltConnection.processNextBatch(DefaultBoltConnection.java:177) ~[neo4j-bolt-4.4.3.jar:4.4.3]
	at org.neo4j.bolt.runtime.scheduling.ExecutorBoltScheduler.executeBatch(ExecutorBoltScheduler.java:257) ~[neo4j-bolt-4.4.3.jar:4.4.3]
	at org.neo4j.bolt.runtime.scheduling.ExecutorBoltScheduler.lambda$scheduleBatchOrHandleError$3(ExecutorBoltScheduler.java:240) ~[neo4j-bolt-4.4.3.jar:4.4.3]
	at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [netty-common-4.1.69.Final.jar:4.1.69.Final]
	at java.lang.Thread.run(Thread.java:829) [?:?]

```
After debugging a little bit I found that this part of the library causes the error
```php
   /**
     * Send DISCARD message
     * The DISCARD message requests that the remainder of the result stream should be thrown away.
     *
     * @link https://7687.org/bolt/bolt-protocol-message-specification-4.html#request-message---discard
     * @param mixed ...$args
     * @return array
     * @throws Exception
     */
    public function discard(...$args): array
    {
        if (count($args) == 0)
            $args[0] = ['n' => -1];
        elseif (!array_key_exists('n', $args[0]))
            $args[0]['n'] = -1;

        $this->write($this->packer->pack(0x2F, $args[0]));
        $message = $this->read($signature);

        if ($signature == self::FAILURE) {
            throw new MessageException($message['message'], $message['code']);
        }

        if ($signature == self::IGNORED) {
            throw new IgnoredException('DISCARD message IGNORED. Server in FAILED or INTERRUPTED state.');
        }

        return $message;
    }
```
The server responds with a failure when the library tried to send a DISCARD message, I'm not sure if I have something wrong in my neo4j configuration or what could be causing this issue.
Worth noting that the query doesn't get completed, so the update doesn't take effect, although there is a commit right there, and although it is logged in the query logs that it was a success, but when I checked the data on the graph the update didn't take effect.

**Timout**
I'm noticing different behaviors for different scenarios when it comes to timeout, I just want to confirm if they are intended or not, as I think some of them might not be intended, so here are the behaviors:

- A query that **doesn't** return and with **low** timeout value (like 10)
```php
$result = $client->run("MATCH (n:Node) SET n.testing = 'hello' WITH * CALL apoc.util.sleep(2000000)");
```
This throws a `Bolt\error\ConnectionTimeoutException` exception and prints an error `Connection timeout reached after 10 seconds.` which is expected and the behavior I thought all scenarios would do.

- A query that **doesn't** return and with **high** timeout value (like 150)
```php
$result = $client->run("MATCH (n:Node) SET n.testing = 'hello' WITH * CALL apoc.util.sleep(2000000)");
$output->writeln("Finished");
```
This doesn't throw any errors on the run step, it proceeds and prints `Finished`, then the library tries to send a DISCARD message to the server, but the server responds with a null message, so the following error is produced
```
In MessageException.php line 32:
                                                                                                                                                                                                           
  Type error: Argument 1 passed to Bolt\error\MessageException::__construct() must be of the type string, null given, called in /var/www/udg.zlien.com/vendor/stefanak-michal/bolt/src/protocol/V4.php on  
   line 94
```
And this error is produced on neo4j
```
2022-03-16 19:09:53.322+0000 ERROR [o.n.b.r.s.i.ErrorReporter] Client triggered an unexpected error [Neo.DatabaseError.General.UnknownError]: null, reference 76c15d57-68ae-47a5-8591-ee67af056a0e.
2022-03-16 19:09:53.322+0000 ERROR [o.n.b.r.s.i.ErrorReporter] Client triggered an unexpected error [Neo.DatabaseError.General.UnknownError]: null, reference 76c15d57-68ae-47a5-8591-ee67af056a0e.
java.lang.NullPointerException: null
	at java.util.concurrent.ConcurrentHashMap.get(ConcurrentHashMap.java:936) ~[?:?]
	at org.neo4j.bolt.transaction.StatementProcessorTxManager.retrieveTx(StatementProcessorTxManager.java:200) ~[neo4j-bolt-4.4.3.jar:4.4.3]
	at org.neo4j.bolt.transaction.StatementProcessorTxManager.streamResults(StatementProcessorTxManager.java:246) ~[neo4j-bolt-4.4.3.jar:4.4.3]
	at org.neo4j.bolt.transaction.StatementProcessorTxManager.discardData(StatementProcessorTxManager.java:118) ~[neo4j-bolt-4.4.3.jar:4.4.3]
	at org.neo4j.bolt.v4.runtime.AutoCommitState.processStreamDiscardResultMessage(AutoCommitState.java:59) ~[neo4j-bolt-4.4.3.jar:4.4.3]
	at org.neo4j.bolt.v4.runtime.AbstractStreamingState.processUnsafe(AbstractStreamingState.java:56) ~[neo4j-bolt-4.4.3.jar:4.4.3]
	at org.neo4j.bolt.v3.runtime.FailSafeBoltStateMachineState.process(FailSafeBoltStateMachineState.java:48) ~[neo4j-bolt-4.4.3.jar:4.4.3]
	at org.neo4j.bolt.runtime.statemachine.impl.AbstractBoltStateMachine.nextState(AbstractBoltStateMachine.java:154) ~[neo4j-bolt-4.4.3.jar:4.4.3]
	at org.neo4j.bolt.runtime.statemachine.impl.AbstractBoltStateMachine.process(AbstractBoltStateMachine.java:102) ~[neo4j-bolt-4.4.3.jar:4.4.3]
	at org.neo4j.bolt.messaging.BoltRequestMessageReader.lambda$doRead$1(BoltRequestMessageReader.java:93) ~[neo4j-bolt-4.4.3.jar:4.4.3]
	at org.neo4j.bolt.runtime.DefaultBoltConnection.lambda$enqueue$0(DefaultBoltConnection.java:156) ~[neo4j-bolt-4.4.3.jar:4.4.3]
	at org.neo4j.bolt.runtime.DefaultBoltConnection.processNextBatchInternal(DefaultBoltConnection.java:252) ~[neo4j-bolt-4.4.3.jar:4.4.3]
	at org.neo4j.bolt.runtime.DefaultBoltConnection.processNextBatch(DefaultBoltConnection.java:187) ~[neo4j-bolt-4.4.3.jar:4.4.3]
	at org.neo4j.bolt.runtime.DefaultBoltConnection.processNextBatch(DefaultBoltConnection.java:177) ~[neo4j-bolt-4.4.3.jar:4.4.3]
	at org.neo4j.bolt.runtime.scheduling.ExecutorBoltScheduler.executeBatch(ExecutorBoltScheduler.java:257) ~[neo4j-bolt-4.4.3.jar:4.4.3]
	at org.neo4j.bolt.runtime.scheduling.ExecutorBoltScheduler.lambda$scheduleBatchOrHandleError$3(ExecutorBoltScheduler.java:240) ~[neo4j-bolt-4.4.3.jar:4.4.3]
	at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [netty-common-4.1.69.Final.jar:4.1.69.Final]
	at java.lang.Thread.run(Thread.java:829) [?:?]
```
I don't believe this behavior is intended at all.

- A query that **does** return and with **low** timeout value (like 10)
```php
$result = $client->run("MATCH (n:Node) SET n.testing = 'hello' WITH * CALL apoc.util.sleep(2000000) RETURN n");
dump($result->toArray());
```
This throws a `Bolt\error\MessageException` exception with the message
```
The transaction has been terminated. Retry your operation in a new transaction, and you should see a successful result. The transaction has not completed within the specified timeout (dbms.transaction.timeout). You may want to retry with a longer timeout.  (Neo.ClientError.Transaction.TransactionTimedOut)
```
I think this behavior is intended, the error message is the one sent from the server so it is descriptive but the exception type is not `ConnectionTimeoutException` in this case (which I'm not sure if that is intended or not).
Note that neo4j logs the following in this case
```
2022-03-16 19:24:41.309+0000 WARN  [o.n.k.i.a.t.m.TransactionMonitor] Transaction QueryRouterTransaction[id=557,clientAddress=172.22.0.9:34622,username=neo4j] timeout.
2022-03-16 19:24:41.309+0000 WARN  [o.n.k.i.a.t.m.TransactionMonitor] Transaction QueryRouterTransaction[id=558,clientAddress=172.22.0.9:34624,username=neo4j] timeout.
2022-03-16 19:24:41.325+0000 WARN  [o.n.b.t.p.HouseKeeper] Fatal error occurred when handling a client connection, remote peer unexpectedly closed connection: [id: 0x4cdd3561, L:/172.22.0.7:7687 - R:/172.22.0.9:34624]
```

- A query that **does** return and with **high** timeout value (like 150)
```php
$result = $client->run("MATCH (n:Node) SET n.testing = 'hello' WITH * CALL apoc.util.sleep(2000000) RETURN n");
dump($result->toArray());
```
This throws a `Bolt\error\IgnoredException` exception with the message
```
DISCARD message IGNORED. Server in FAILED or INTERRUPTED state.
```
I don't think this behavior is intended, the exception is produced when the library tried to send the DISCARD message to the server, but the server ignores it.

**NOTE**: for all the scenarios, the timeout value is respected, so no scenario finished before the timeout value was reached.

Apologies for the very long write-down :pray: Any help/clarification on these issues would be extremely appreciated :pray: 

Client version: 2.5.2
Neo4j version: 4.4.3

**Comment by transistive** [comment]
Hello @noureldin-khaled,

No need to apologize at all! I am flattered and grateful you took the time to write such an excellent and comprehensive issue. I'll try and fix them today.

The dump bug is an interesting side effect. I do indeed a bit of reference magic to allow for lazy evaluation. I'll try and look for a solution that doesn't break dumping, and presumably, serialization.

**Comment by transistive** [comment]
I will also take a time to break down the issue in multiple, smaller ones

**Comment by transistive** [comment]
Hello @noureldin-khaled,

Thank you for your patience. It's been quite busy, and I am on what is looking to be just a semi holiday. I hope to be able to fix all the issues soon. Issue #128 is already fixed, and I am working on #127 now

**Comment by noureldin-khaled** [comment]
Sorry for the reminder again :sweat_smile: any updates on these issues?

**Comment by transistive** [comment]
Hello @noureldin-khaled,

Apologies for the wait. I will at least merge two of the three sub issues today. The final one is quite complex and requires a small redesign which I am working on right now.

Kind regards,

Ghle

**Comment by transistive** [comment]
Closed by #131, #130, #128 

It is available in dev-master right now, and should be released before monday under version 2.3.0
